Add pill to inform about card requirement when OC is enabled

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-645
Base PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4540

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR adds a new pill to notify merchants when the Optimized Checkout is enabled that the payment method cannot be disabled. Currently, it is required for OC to work correctly.

Preview:
<img width="752" height="208" alt="Screenshot 2025-07-30 at 12 13 31" src="https://github.com/user-attachments/assets/c1395e34-8015-4619-82fa-020580c2de51" />

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout to this branch on your test environment (`add/pill-to-inform-about-card-requirement-when-oc-is-enabled`)
- Connect your Stripe account
- Make OC available (you can use the `wc_stripe_is_optimized_checkout_available` filter for that)
- Enable OC
- Go to the payment methods list screen
- Check if you can see the new pill on top of the credit card method (compare with the preview above)

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Blocking disabling of cards when OC is enabled

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Add new pill to inform about the card requirement when OC is enabled

wjrosa committed:
Block the checkbox as well

wjrosa committed:
Merge branch 'develop' into fix/block-disabling-of-cards-when-oc-is-enabled

wjrosa committed:
Refactor logic to force enablement during init

wjrosa committed:
Unit tests

wjrosa committed:
Fix tests

diegocurbelo commented on client/components/payment-method-unavailable-due-conflict-pill/index.js:
```diff
@@ -48,25 +45,22 @@ const IconComponent = ( { children, ...props } ) => (
 );
 
-const PaymentMethodUnavailableDueConflictPill = ( { id, label } ) => {
+const PaymentMethodRequiredForOCPill = ( { id, label } ) => {
```
I think the changes in this file were intended for the new file `client/components/payment-method-required-for-oc-pill/index.js`

diegocurbelo commented on client/components/payment-method-required-for-oc-pill/index.js:
```diff
@@ -0,0 +1,96 @@
+	</IconWrapper>
+);
+
+const PaymentMethodUnavailableDueConflictPill = ( { id, label } ) => {
```
I think this file got mixed with `client/components/payment-method-unavailable-due-conflict-pill/index.js`, the changes in that file should be moved here.

diegocurbelo reviewed (changes requested):
It seems that the contents of the existing `client/components/payment-method-unavailable-due-conflict-pill/index.js` and the new `client/payment-method-required-for-oc-pill/index.js` got swapped, which breaks the Affirm/Klarna not available due conflict pill.

Also, the disabled checkbox styles might be confusing. It looks the same when the payment method is `enabled and can't be disabled` as when they are `disabled but can't be enabled`, even with the new `Required` pill, they might incorrectly think that Cards is not enabled.
I don't think we have an "active but disabled" style for the checkbox, so maybe we can leave it enabled but make it not interactable?

<img width="758" height="567" alt="Screenshot_2025-07-31_at_10_46_33" src="https://github.com/user-attachments/assets/e39b2d57-7841-484a-af78-787177b7ebe9" />

wjrosa committed:
Merge branch 'fix/block-disabling-of-cards-when-oc-is-enabled' into add/pill-to-inform-about-card-requirement-when-oc-is-enabled

wjrosa committed:
Fix payment method pills swapping

wjrosa committed:
Testing custom checked+disabled style

wjrosa commented:
Thanks for the review again, Diego!

> It seems that the contents of the existing client/components/payment-method-unavailable-due-conflict-pill/index.js and the new client/payment-method-required-for-oc-pill/index.js got swapped, which breaks the Affirm/Klarna not available due conflict pill.

Oops! Sorry about that 🤦 . I should not code during meetings. I have fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4541/commits/a3cde9319277ea3fa27469762aa499c35cd56edb

> I don't think we have an "active but disabled" style for the checkbox, so maybe we can leave it enabled but make it not interactable?

I tried to create a custom style for it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4541/commits/a2d8f685a11302d676eef1dc721b391fa0586c2c. Preview:
<img width="766" height="217" alt="Screenshot 2025-07-31 at 11 50 51" src="https://github.com/user-attachments/assets/995f2895-9481-42b7-8ba3-bad861e723d6" />

But I think we should not diverge from the current standards, so I reverted it, and now it looks like a regular checked checkbox (but not interactive as you suggested).

wjrosa committed:
Remove custom checkbox style

wjrosa committed:
Merge branch 'develop' into add/pill-to-inform-about-card-requirement-when-oc-is-enabled

diegocurbelo commented on client/settings/general-settings-section/payment-method-checkbox.js:
```diff
@@ -44,9 +45,17 @@ const PaymentMethodCheckbox = ( {
 	] = useEnabledPaymentMethodIds();
 	const [ , setIsStripeEnabled ] = useIsStripeEnabled();
 	const { isUpeEnabled } = useContext( UpeToggleContext );
+	let checked;
+	if ( disabled ) {
+		checked = false;
+	} else if ( disabledButChecked ) {
+		checked = true;
+	} else {
+		checked = enabledPaymentMethods.includes( id );
+	}
```
[nit] This is more of personal preference... feel free to ignore it!
```suggestion
	const checked = !disabled && ( disabledButChecked || enabledPaymentMethods.includes( id ) );
```

diegocurbelo reviewed (approved):
Thanks for the change @wjrosa, it looks great! 🚢 

I like your custom "dimmed" style, but I'm also fine using the normal "checked" style, both make it clear that the payment method is enabled.

malithsen commented on tests/phpunit/WC_Stripe_Test.php:
```diff
@@ -6,6 +6,7 @@
 use WC_Stripe_Helper;
 use WC_Stripe_Payment_Methods;
 use WC_Stripe_UPE_Payment_Gateway;
+use WooCommerce\Stripe\Tests\Helpers\OC_Test_Helper;
```
Is this import required?

wjrosa replied:
Not really. Removed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4541/commits/4b5ddd3a168710e196c2b98ab69f565c42e9e797

malithsen reviewed (changes requested):
Nice work. I noticed that a refresh is needed for the checkbox to be disabled on the payment method list. Can we fix that?

https://github.com/user-attachments/assets/48d32f54-484b-4588-acb2-c6894462b046

malithsen reviewed (approved):
Actually, I'll change my previous review to approve this as the immediate update of the state is not critical for this particular checkbox but it would still be nice to fix it.

wjrosa committed:
Merge branch 'develop' into add/pill-to-inform-about-card-requirement-when-oc-is-enabled

wjrosa committed:
Removing unnecessary import

wjrosa committed:
Removing the need for a page refresh

wjrosa commented:
Thanks, Malith! I have removed the need for a refresh in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4541/commits/531d806118bfe8626c0fe3b9b300a28ccb886727

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Update client/settings/general-settings-section/payment-method-checkbox.js

Co-authored-by: Diego Curbelo <diego@curbelo.com>

wjrosa committed:
Fix lint issue
